### ISSUE TITLE
Prevent Quartz jobs from running during Liquibase migrations

### ISSUE THREAD

**Issue by rhodebump** [openboxes/openboxes#26]
When I startup a fresh install of openboxes with mysql, I see the following stack trace.

2014-10-23 18:58:11,167 [main] INFO  liquibase  - Changeset 0.5.x/add-location_role.xml::1355247924098-3::Thoughtworks (generated)::(MD5Sum: d73fd151db672ad79ad45cbf9a77d69)
2014-10-23 18:58:11,185 [main] INFO  liquibase  - Changeset 0.5.x/add-location_role.xml::1355247924098-4::Thoughtworks (generated)::(MD5Sum: c26c78db85d870f6f7e573fe53c531d2)
2014-10-23 18:58:11,205 [main] INFO  liquibase  - Changeset 0.5.x/add-product-fields.xml::1358133447150-1::jmiranda (generated)::(MD5Sum: 33beb26ecbb06277e59180e83881151d)
2014-10-23 18:58:12,253 [DefaultPluginScheduler_Worker-1] WARN  util.JDBCExceptionReporter  - SQL Error: 1054, SQLState: 42S22
2014-10-23 18:58:12,257 [DefaultPluginScheduler_Worker-1] ERROR util.JDBCExceptionReporter  - Unknown column 'product0_.active' in 'field list'
2014-10-23 18:58:12,354 [DefaultPluginScheduler_Worker-1] ERROR quartz2.JobErrorLoggerListener  - Exception thrown in job:org.pih.warehouse.jobs.AssignIdentifierJob
org.quartz.JobExecutionException: org.springframework.dao.InvalidDataAccessResourceUsageException: could not execute query; SQL [select product0_.id as id19_, product0_.version as version19_, product0_.active as active19_, product0_.brand_name as brand4_19_, product0_.category_id as category5_19_, product0_.cold_chain as cold6_19_, product0_.controlled_substance as controlled7_19_, product0_.created_by_id as created8_19_, product0_.date_created as date9_19_, product0_.default_uom_id as default10_19_, product0_.description as descrip11_19_, product0_.essential as essential19_, product0_.hazardous_material as hazardous13_19_, product0_.last_updated as last14_19_, product0_.lot_control as lot15_19_, product0_.manufacturer as manufac16_19_, product0_.manufacturer_code as manufac17_19_, product0_.manufacturer_name as manufac18_19_, product0_.model_number as model19_19_, product0_.name as name19_, product0_.ndc as ndc19_, product0_.package_size as package22_19_, product0_.price_per_unit as price23_19_, product0_.product_code as product24_19_, product0_.reconditioned as recondi25_19_, product0_.serialized as serialized19_, product0_.unit_of_measure as unit27_19_, product0_.upc as upc19_, product0_.updated_by_id as updated29_19_, product0_.vendor as vendor19_, product0_.vendor_code as vendor31_19_, product0_.vendor_name as vendor32_19_ from product product0_ where product0_.product_code is null or product0_.product_code='']; nested exception is org.hibernate.exception.SQLGrammarException: could not execute query [See nested exception: org.springframework.dao.InvalidDataAccessResourceUsageException: could not execute query; SQL [select product0_.id as id19_, product0_.version as version19_, product0_.active as active19_, product0_.brand_name as brand4_19_, product0_.category_id as category5_19_, product0_.cold_chain as cold6_19_, product0_.controlled_substance as controlled7_19_, product0_.created_by_id as created8_19_, product0_.date_created as date9_19_, product0_.default_uom_id as default10_19_, product0_.description as descrip11_19_, product0_.essential as essential19_, product0_.hazardous_material as hazardous13_19_, product0_.last_updated as last14_19_, product0_.lot_control as lot15_19_, product0_.manufacturer as manufac16_19_, product0_.manufacturer_code as manufac17_19_, product0_.manufacturer_name as manufac18_19_, product0_.model_number as model19_19_, product0_.name as name19_, product0_.ndc as ndc19_, product0_.package_size as package22_19_, product0_.price_per_unit as price23_19_, product0_.product_code as product24_19_, product0_.reconditioned as recondi25_19_, product0_.serialized as serialized19_, product0_.unit_of_measure as unit27_19_, product0_.upc as upc19_, product0_.updated_by_id as updated29_19_, product0_.vendor as vendor19_, product0_.vendor_code as vendor31_19_, product0_.vendor_name as vendor32_19_ from product product0_ where product0_.product_code is null or product0_.product_code='']; nested exception is org.hibernate.exception.SQLGrammarException: could not execute query]
    at grails.plugin.quartz2.GrailsArtefactJob.execute(GrailsArtefactJob.java:66)
    at org.quartz.core.JobRunShell.run(JobRunShell.java:213)
    at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:557)
Caused by: org.springframework.dao.InvalidDataAccessResourceUsageException: could not execute query; SQL [select product0_.id as id19_, product0_.version as version19_, product0_.active as active19_, product0_.brand_name as brand4_19_, product0_.category_id as category5_19_, product0_.cold_chain as cold6_19_, product0_.controlled_substance as controlled7_19_, product0_.created_by_id as created8_19_, product0_.date_created as date9_19_, product0_.default_uom_id as default10_19_, product0_.description as descrip11_19_, product0_.essential as essential19_, product0_.hazardous_material as hazardous13_19_, product0_.last_updated as last14_19_, product0_.lot_control as lot15_19_, product0_.manufacturer as manufac16_19_, product0_.manufacturer_code as manufac17_19_, product0_.manufacturer_name as manufac18_19_, product0_.model_number as model19_19_, product0_.name as name19_, product0_.ndc as ndc19_, product0_.package_size as package22_19_, product0_.price_per_unit as price23_19_, product0_.product_code as product24_19_, product0_.reconditioned as recondi25_19_, product0_.serialized as serialized19_, product0_.unit_of_measure as unit27_19_, product0_.upc as upc19_, product0_.updated_by_id as updated29_19_, product0_.vendor as vendor19_, product0_.vendor_code as vendor31_19_, product0_.vendor_name as vendor32_19_ from product product0_ where product0_.product_code is null or product0_.product_code='']; nested exception is org.hibernate.exception.SQLGrammarException: could not execute query
    at org.springframework.orm.hibernate3.SessionFactoryUtils.convertHibernateAccessException(SessionFactoryUtils.java:629)
    at org.springframework.orm.hibernate3.HibernateAccessor.convertHibernateAccessException(HibernateAccessor.java:412)
    at org.springframework.orm.hibernate3.HibernateTemplate.doExecute(HibernateTemplate.java:411)
    at org.springframework.orm.hibernate3.HibernateTemplate.executeFind(HibernateTemplate.java:343)
    at org.codehaus.groovy.grails.orm.hibernate.metaclass.FindAllPersistentMethod.doInvokeInternal(FindAllPersistentMethod.java:122)
    at org.codehaus.groovy.grails.orm.hibernate.metaclass.AbstractStaticPersistentMethod.invoke(AbstractStaticPersistentMethod.java:64)
    at org.codehaus.groovy.grails.orm.hibernate.metaclass.AbstractStaticPersistentMethod.invoke(AbstractStaticPersistentMethod.java:57)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:606)
    at org.codehaus.groovy.runtime.callsite.PojoMetaMethodSite$PojoCachedMethodSite.invoke(PojoMetaMethodSite.java:188)
    at org.codehaus.groovy.runtime.callsite.PojoMetaMethodSite.call(PojoMetaMethodSite.java:52)
    at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:40)
    at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:116)
    at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:132)
    at org.codehaus.groovy.grails.plugins.orm.hibernate.HibernatePluginSupport$_enhanceSessionFactory_closure11.doCall(HibernatePluginSupport.groovy:371)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:606)
    at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:90)
    at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:233)
    at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1058)
    at groovy.lang.ExpandoMetaClass.invokeMethod(ExpandoMetaClass.java:1070)
    at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:886)
    at groovy.lang.Closure.call(Closure.java:282)
    at org.codehaus.groovy.runtime.metaclass.ClosureStaticMetaMethod.invoke(ClosureStaticMetaMethod.java:59)
    at org.codehaus.groovy.runtime.callsite.StaticMetaMethodSite$StaticMetaMethodSiteNoUnwrapNoCoerce.invoke(StaticMetaMethodSite.java:148)
    at org.codehaus.groovy.runtime.callsite.StaticMetaMethodSite.call(StaticMetaMethodSite.java:88)
    at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:40)
    at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:116)
    at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:124)
    at org.pih.warehouse.core.IdentifierService.assignProductIdentifiers(IdentifierService.groovy:138)
    at org.pih.warehouse.core.IdentifierService$$FastClassByCGLIB$$5c191f1.invoke(<generated>)
    at net.sf.cglib.proxy.MethodProxy.invoke(MethodProxy.java:149)
    at org.springframework.aop.framework.Cglib2AopProxy$CglibMethodInvocation.invokeJoinpoint(Cglib2AopProxy.java:688)
    at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150)
    at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:110)
    at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
    at org.springframework.aop.framework.Cglib2AopProxy$DynamicAdvisedInterceptor.intercept(Cglib2AopProxy.java:621)
    at org.pih.warehouse.core.IdentifierService$$EnhancerByCGLIB$$ec753bd.assignProductIdentifiers(<generated>)
    at org.pih.warehouse.core.IdentifierService$assignProductIdentifiers.call(Unknown Source)
    at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:40)
    at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:116)
    at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:120)
    at org.pih.warehouse.jobs.AssignIdentifierJob.execute(AssignIdentifierJob.groovy:15)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:606)
    at grails.plugin.quartz2.GrailsArtefactJob.execute(GrailsArtefactJob.java:59)
    ... 2 more
Caused by: org.hibernate.exception.SQLGrammarException: could not execute query
    at org.hibernate.exception.SQLStateConverter.convert(SQLStateConverter.java:90)
    at org.hibernate.exception.JDBCExceptionHelper.convert(JDBCExceptionHelper.java:66)
    at org.hibernate.loader.Loader.doList(Loader.java:2231)
    at org.hibernate.loader.Loader.listIgnoreQueryCache(Loader.java:2125)
    at org.hibernate.loader.Loader.list(Loader.java:2120)
    at org.hibernate.loader.hql.QueryLoader.list(QueryLoader.java:401)
    at org.hibernate.hql.ast.QueryTranslatorImpl.list(QueryTranslatorImpl.java:361)
    at org.hibernate.engine.query.HQLQueryPlan.performList(HQLQueryPlan.java:196)
    at org.hibernate.impl.SessionImpl.list(SessionImpl.java:1148)
    at org.hibernate.impl.QueryImpl.list(QueryImpl.java:102)
    at org.codehaus.groovy.grails.orm.hibernate.metaclass.FindAllPersistentMethod$1.doInHibernate(FindAllPersistentMethod.java:186)
    at org.springframework.orm.hibernate3.HibernateTemplate.doExecute(HibernateTemplate.java:406)
    ... 51 more
Caused by: com.mysql.jdbc.exceptions.jdbc4.MySQLSyntaxErrorException: Unknown column 'product0_.active' in 'field list'
    at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
    at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:57)
    at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
    at java.lang.reflect.Constructor.newInstance(Constructor.java:526)
    at com.mysql.jdbc.Util.handleNewInstance(Util.java:411)
    at com.mysql.jdbc.Util.getInstance(Util.java:386)
    at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:1054)
    at com.mysql.jdbc.MysqlIO.checkErrorPacket(MysqlIO.java:4190)
    at com.mysql.jdbc.MysqlIO.checkErrorPacket(MysqlIO.java:4122)
    at com.mysql.jdbc.MysqlIO.sendCommand(MysqlIO.java:2570)
    at com.mysql.jdbc.MysqlIO.sqlQueryDirect(MysqlIO.java:2731)
    at com.mysql.jdbc.ConnectionImpl.execSQL(ConnectionImpl.java:2818)
    at com.mysql.jdbc.PreparedStatement.executeInternal(PreparedStatement.java:2157)
    at com.mysql.jdbc.PreparedStatement.executeQuery(PreparedStatement.java:2324)
    at com.mchange.v2.c3p0.impl.NewProxyPreparedStatement.executeQuery(NewProxyPreparedStatement.java:76)
    at org.hibernate.jdbc.AbstractBatcher.getResultSet(AbstractBatcher.java:208)
    at org.hibernate.loader.Loader.getResultSet(Loader.java:1808)
    at org.hibernate.loader.Loader.doQuery(Loader.java:697)
    at org.hibernate.loader.Loader.doQueryAndInitializeNonLazyCollections(Loader.java:259)
    at org.hibernate.loader.Loader.doList(Loader.java:2228)
    ... 60 more
2014-10-23 18:58:12,650 [main] INFO  liquibase  - Changeset 0.5.x/add-product-fields.xml::1358133447150-2::jmiranda (generated)::(MD5Sum: f8e177c378a88dd5149b793b1c75b0cd)
2014-10-23 18:58:14,118 [main] INFO  liquibase  - Changeset 0.5.x/add-product-fields.xml::1358133447150-3::jmiranda (generated)::(MD5Sum: 16d53e24311682f3cb74ddaaf31284a9)
2014-10-23 18:58:15,531 [main] INFO  liquibase  - Changeset 0.5.x/add-product-fields.xml::1358133447150-4::jmiranda (generated)::(MD5Sum: bb358abdda5cc8feb715d7376cb022c4)
2014-10-23 18:58:16,923 [main] INFO  liquibase  - Changeset 0.5.x/add-product-fields.xml::1358133447150-5::jmiranda (generated)::(MD5Sum: 49f872d5d8b0bbe6fd8ec12f19e95)
2014-10-23 18:58:18,355 [main] INFO  liquibase  - Changeset 0.5.x/add-product-fields.xml::1358133447150-6::jmiranda (generated)::(MD5Sum: 675274e5a1b5ed2d219f3d28578b15)
2014-10-23 18:58:19,771 [main] INFO  liquibase  - Changeset 0.5.x/add-product-fields.xml::1358133447150-7::jmiranda (generated)::(MD5Sum: 6ecb28d5b63a5496d5f17b733b1cbf94)
2014-10-23 18:58:21,195 [main] INFO  liquibase  - Changeset 0.5.x/add-inventory-level-max-quantity.xml::1358886163959-1::jmiranda (generated)::(MD5Sum: 3b4ecc54b35f34fc39c6cd3274b16c52)
2014-10-23 18:58:22,515 [main] INFO  liquibase  - Changeset 0.5.x/add-inventory-level-bin-location.xml::1358886163959-2::jmiranda (generated)::(MD5Sum: 148c02443c34158293994497e59a61c)
2014-10-23 18:58:23,933 [main] INFO  liquibase  - Changeset 0.5.x/add-transaction-number.xml::1359144266832-1::jmiranda (generated)::(MD5Sum: 23459577e8bf5ef5866dc7cff3ae6c6)

**Comment by jmiranda** [comment]
After doing some research, the "best" solution is to set the Quartz job's startDelay to a value greater than the time it takes to run through the entire Liquibase process from scratch.  For a new installation, that might be two minutes.  However for a system with lots of data that is being upgraded to the latest version of OpenBoxes it might take considerably longer -- especially if there indexes created on tables with lots of data.  I'll run through the "from scratch" scenario and come up with a good estimate.  

However, it would be better to delay the start until Liquibase has actually completed it's work not just when we *think* it should be done.  In other words, all Quartz jobs that rely on a domain class (e.g. have a database dependency) should implement an is-database-migration-completed interface and should delay their execution until Liquibase has completed all migrations.  

Unfortunately, this doesn't seem like it would be a simple solution, so for now I'm going to go with the more hacky "wait a reasonable amount of time before starting" solution.

**Comment by jmiranda** [comment]
Just realized that the solution I proposed years ago should be pretty simple to implement. We can just create an abstract class called DatabaseDependencyJob that postpones execution until there are no DATABASECHANGELOGLOCK records. It's couples Quartz and Liquibase but I'm ok with that for the time being if it means we don't have to see these ugly stacktraces on initial install.

**Comment by jmiranda** [comment]
I thought this was fixed, but need to look into another solution since the current solution might be causing threads to get stuck. The alternative solution would be to disable autoStartup and start the scheduler once the database migration have been completed.  
```
grails.plugin.quartz2.autoStartup = false
```

**Comment by jmiranda** [comment]
The thread stuck issue no longer seems to be a problem on the demo server so I'm going to close this. A possible workaround would be to increase the thread count allocated to Quartz. 
```
org {
        quartz{
                //anything here will get merged into the quartz.properties so you don't need another file
                threadPool.class = 'org.quartz.simpl.SimpleThreadPool'
                threadPool.threadCount = 100
                threadPool.threadsInheritContextClassLoaderOfInitializingThread = true
                jobStore.class = 'org.quartz.simpl.RAMJobStore'
        }
}
```
The OBGM (Upgrade to Grails 3) project solves this by starting the scheduler AFTER the database migrations have completed.
https://github.com/openboxes/openboxes/blob/feature/upgrade-to-grails-3.3.10/grails-app/init/org/pih/warehouse/BootStrap.groovy#L90
